src/lib/tls/tls13/tls_extensions_psk.cpp

```diff
@@ -92,32 +105,32 @@ class Client_PSK {
 
 class Server_PSK {
    public:
-      Server_PSK(uint16_t id) : m_selected_identity(id), m_session_to_resume(std::nullopt) {}
+      Server_PSK(uint16_t id) : m_selected_identity(id), m_session_to_resume_or_psk(std::monostate()) {}
 
-      Server_PSK(uint16_t id, Session session) : m_selected_identity(id), m_session_to_resume(std::move(session)) {}
+      Server_PSK(uint16_t id, Session session) :
+            m_selected_identity(id), m_session_to_resume_or_psk(std::move(session)) {}
+
+      Server_PSK(uint16_t id, ExternalPSK psk) : m_selected_identity(id), m_session_to_resume_or_psk(std::move(psk)) {}
 
       uint16_t selected_identity() const { return m_selected_identity; }
 
-      Session take_session_to_resume() {
-         BOTAN_STATE_CHECK(m_session_to_resume.has_value());
-         Session s = std::move(m_session_to_resume.value());
-         m_session_to_resume = std::nullopt;
-         return s;
+      std::variant<std::monostate, Session, ExternalPSK> take_session_to_resume_or_psk() {
+         BOTAN_STATE_CHECK(!std::holds_alternative<std::monostate>(m_session_to_resume_or_psk));
+         return std::exchange(m_session_to_resume_or_psk, std::monostate());
```

Comment:
I fully agree that `std::monostate` is a hack and shouldn't leak into the public interface. We use it because it is easier to use than the suggested `std::optional<std::variant<>>` combination. The best we can do in terms of pattern matching is the `overloaded{}` helper, and that becomes more compact using `std::monostate` instead of `std::optional<>`.

All that to say: This particular method is an internal class (`Server_PSK` is defined in an anonymous namespace). The public interface already decomposes this and returns `std::variant<Session, ExternalPSK>` as you suggest. See `PSK::take_session_to_resume_or_psk()` in line 360ish.

Reply:
I saw that this is already cleaned up in `PSK::take_session_to_resume_or_psk`, which is the important one - it just feels like we could avoid the possible assertion there if the `ServerPSK` return type was tighter.

> We use it because it is easier to use than the suggested std::optional<std::variant<>> combination. The best we can do in terms of pattern matching is the overloaded{} helper

Rust has really spoiled me in terms of this kind of thing being easy :/

Resolving because this is really not that big a deal.